### PR TITLE
Introduce buttons for hot reload

### PR DESCRIPTION
**Pull request by DaveShuckerow**
image

I need to move the banner to the bottom to work around #1176.

### CODE REVIEW

File: packages/devtools_app/lib/src/flutter/app.dart [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 import 'dart:async';
 
+import 'package:devtools_app/src/ui/flutter/service_extension_widgets.dart';
```
Comment (jacob314) [inline packages/devtools_app/lib/src/flutter/app.dart]: make this a relative path. There is a quickfix for that.

File: packages/devtools_app/lib/src/flutter/app.dart [hunk 2]
```diff
@@ -41,7 +42,13 @@ class DevToolsAppState extends State<DevToolsApp> {
     final path = uri.path;
 
     if (_routes.containsKey(path)) {
-      return MaterialPageRoute(settings: settings, builder: _routes[path]);
+      var builder = _routes[path];
+      assert(() {
+        builder =
+            (context) => _AlternateCheckedModeBanner(builder: _routes[path]);
```
Comment (jacob314) [inline packages/devtools_app/lib/src/flutter/app.dart]: we don't need the checked mode banner. Go ahead and just hide it by default instead.

File: packages/devtools_app/lib/src/ui/flutter/service_extension_widgets.dart [hunk 3]
```diff
@@ -227,16 +230,50 @@ class _RegisteredServiceExtensionButtonState
...
-        if (widget.showTitle)
-          Padding(
-            padding: const EdgeInsets.only(left: 8.0),
-            child: Text(widget.serviceDescription.title),
-          ),
-      ]),
+    return InkWell(
+      onTap: _click,
+      child: Container(
+        constraints: const BoxConstraints(minWidth: 48.0, minHeight: 48.0),
```
Comment (jacob314) [inline packages/devtools_app/lib/src/ui/flutter/service_extension_widgets.dart]: why not width 48, height 48? something will be badly wrong if the icons are more than 48 pixels high

File: packages/devtools_app/lib/src/ui/flutter/service_extension_widgets.dart [hunk 4]
```diff
@@ -227,16 +230,50 @@ class _RegisteredServiceExtensionButtonState
...
+
+class HotRestartButton extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return RegisteredServiceExtensionButton(
+      hotRestart,
+      () => serviceManager.performHotRestart(),
+      (arg) {
+        return Scaffold.of(context).showSnackBar(
+          SnackBar(content: Text('Unable to hot restart the app: $arg.')),
```
Comment (jacob314) [inline packages/devtools_app/lib/src/ui/flutter/service_extension_widgets.dart]: we should also display that the hor reload and restart are in progress

File: packages/devtools_app/lib/src/flutter/app.dart [hunk 5]
```diff
@@ -72,6 +79,10 @@ class DevToolsAppState extends State<DevToolsApp> {
               EmptyScreen.logging,
               const InfoScreen(),
             ],
+            actions: [
+              HotReloadButton(),
+              HotRestartButton(),
+            ],
```
Comment (jacob314) [inline packages/devtools_app/lib/src/flutter/app.dart]: add a bit of padding on the end. these icons look a bit close to the right edge of the view.

File: packages/devtools_app/lib/src/ui/flutter/service_extension_widgets.dart [hunk 6]
```diff
@@ -227,16 +230,50 @@ class _RegisteredServiceExtensionButtonState
...
+  }
+}
+
+class HotReloadButton extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return RegisteredServiceExtensionButton(
+      hotReload,
+      () => serviceManager.performHotReload(),
+      (arg) {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/ui/flutter/service_extension_widgets.dart]: These icons look too colorful. make them black and white as in the devtools dart:html app.
Screen Shot 2019-10-21 at 9 14 33 PM

File: packages/devtools_app/lib/src/flutter/app.dart (hunk 2 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/flutter/app.dart]: +1

Comment (DaveShuckerow) [reply packages/devtools_app/lib/src/flutter/app.dart]: It's more than an affordance for understanding that checked mode is slow.  This banner is also a warning that a release has accidentally been built with the wrong mode of the app.  This has saved several teams' release engineers a couple of times when doing releases, even inside Google.

Especially if we look into shipping desktop app builds along with the web app, having the little extra safety will be very helpful.

File: packages/devtools_app/lib/src/ui/flutter/service_extension_widgets.dart (hunk 3 above)
Comment (DaveShuckerow) [reply packages/devtools_app/lib/src/ui/flutter/service_extension_widgets.dart]: It looked like the original intention of this widget was to also show text on the side.

I'm going to rewrite this build method to only show the icon.

File: packages/devtools_app/lib/src/flutter/app.dart (hunk 2 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/app.dart]: Ok lets leave it as is. Coming from a web background it seems unnecessary but maybe it is more important on desktop/mobile where you can't quickly view source to see whether you shipped the wrong files.

File: packages/devtools_app/lib/src/ui/flutter/service_extension_widgets.dart (hunk 3 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/ui/flutter/service_extension_widgets.dart]: sg. Feel free to also leave as is.

File: packages/devtools_app/lib/src/ui/flutter/service_extension_widgets.dart (hunk 6 above)
Comment (DaveShuckerow) [reply packages/devtools_app/lib/src/ui/flutter/service_extension_widgets.dart]: Done.

File: packages/devtools_app/lib/src/flutter/app.dart (hunk 5 above)
Comment (DaveShuckerow) [reply packages/devtools_app/lib/src/flutter/app.dart]: SGTM

File: packages/devtools_app/lib/src/ui/flutter/service_extension_widgets.dart (hunk 4 above)
Comment (DaveShuckerow) [reply packages/devtools_app/lib/src/ui/flutter/service_extension_widgets.dart]: Done.  We'll want to be smart about how we show notifications if we're potentially generating many in a few seconds.  Filed #1242 about this.

File: packages/devtools_app/lib/src/flutter/app.dart (hunk 1 above)
Comment (DaveShuckerow) [reply packages/devtools_app/lib/src/flutter/app.dart]: Done.

File: packages/devtools_app/test/flutter/service_extension_widgets_test.dart [hunk 7]
```diff
@@ -0,0 +1,73 @@
+import 'dart:async';
```
Comment (jacob314) [inline packages/devtools_app/test/flutter/service_extension_widgets_test.dart]: add the standard copyright notice.

File: packages/devtools_app/lib/src/flutter/app.dart (hunk 5 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/flutter/app.dart]: Some padding on the right side would also be nice as the icons look a bit close to the right edge of the view.

**Pull request merged**